Beds 526/Raw DB sdk

Tangui-Bitfly opened the pull request:
* add `db2.BigTableEthRaw` which only purpose is reading from the raw bigtable store
* `db2.BigTableEthRaw` is a `http.RoundTriper`, this allow the caller to use when creating a new `ethclient.Client`
* introduce the `db2.RawStore` to handle all interaction with bigtable satore. There are some duplications here, the aim of this pr is not to refactor everything, it's just to provide an easy to use way of retrieving eth data from bigtable 
* it is used as follow
```go
bt, err := store.NewBigTableWithClient(context.Background(), client, admin, raw)
if err != nil {
	t.Fatal(err)
}

rawStore := NewRawStore(store.Wrap(bt, BlocRawTable, ""))

rpcClient, err := rpc.DialOptions(context.Background(), "http://foo.bar", rpc.WithHTTPClient(&http.Client{
	Transport: NewBigTableEthRaw(rawStore, chainID),
}))
ethClient := ethclient.NewClient(rpcClient)
```

Tangui-Bitfly committed:
chore(ci): update go-ethereum, remove gobitfly/prysm and golangci check

Monika-Bitfly commented on backend/internal/contracts/README.md:
```diff
@@ -0,0 +1,28 @@
+
+- solc is available [here](https://docs.soliditylang.org/en/latest/installing-solidity.html)
+
+- and `abigen` wan be installed through go directly
```
```suggestion
- and `abigen` can be installed through go directly
```

Monika-Bitfly commented on backend/internal/th/ethereum.go:
```diff
@@ -0,0 +1,224 @@
+
+	head, _ := b.Backend.Client().HeaderByNumber(context.Background(), nil)
+	gasPrice := new(big.Int).Add(head.BaseFee, big.NewInt(params.GWei))
+	chainid, _ := b.Backend.Client().ChainID(context.Background())
```
```suggestion
	chainid, err := b.Backend.Client().ChainID(context.Background())
```

Monika-Bitfly commented on backend/internal/th/ethereum.go:
```diff
@@ -0,0 +1,224 @@
+func (b *BlockchainBackend) MakeUnsignedTx(t *testing.T, from common.Address, to *common.Address, value *big.Int, data []byte) *types.Transaction {
+	t.Helper()
+
+	head, _ := b.Backend.Client().HeaderByNumber(context.Background(), nil)
```
```suggestion
	head, err := b.Backend.Client().HeaderByNumber(context.Background(), nil)
```
we should check for any errors coz we will be accessing the BaseFee value to calculate the gas price

Tangui-Bitfly replied:
Since it's only a test file, I though that a simple panic was fine.

Monika-Bitfly commented on backend/internal/th/ethereum.go:
```diff
@@ -0,0 +1,224 @@
+func (b *BlockchainBackend) DeployContract(t *testing.T, contractData []byte) common.Address {
+	t.Helper()
+
+	head, _ := b.Backend.Client().HeaderByNumber(context.Background(), nil)
```
```suggestion
	head, err := b.Backend.Client().HeaderByNumber(context.Background(), nil)
```

Monika-Bitfly commented on backend/internal/th/ethereum.go:
```diff
@@ -0,0 +1,224 @@
+
+	head, _ := b.Backend.Client().HeaderByNumber(context.Background(), nil)
+	gasPrice := new(big.Int).Add(head.BaseFee, big.NewInt(params.GWei))
+	chainid, _ := b.Backend.Client().ChainID(context.Background())
```
```suggestion
	chainid, err := b.Backend.Client().ChainID(context.Background())
```

Monika-Bitfly commented on backend/internal/th/ethereum.go:
```diff
@@ -0,0 +1,224 @@
+	bk := ethsimtracer.NewBackend(genesis, func(nodeConf *node.Config, ethConf *ethconfig.Config) {
+		nodeConf.HTTPModules = append(nodeConf.HTTPModules, "debug", "eth")
+		nodeConf.HTTPHost = "127.0.0.1"
+		nodeConf.HTTPPort = 4242
```
It might be a good idea to move the port value to the top of the file and store it as a constant - it will make it easier to update when needed

Monika-Bitfly commented on backend/pkg/commons/db2/database/bigtable.go:
```diff
@@ -0,0 +1,347 @@
+		return fmt.Errorf("cannot ApplyBulk err: %w", err)
+	}
+	// TODO aggregate errs
+	for _, e := range errs {
```
```suggestion
	var bulkErrs []string
        for _, e := range errs {
            bulkErrs = append(bulkErrs, e.Error())
        }
        if len(bulkErrs) > 0 {
            return fmt.Errorf("cannot ApplyBulk elem, errors: %v", bulkErrs)
        }
```

Monika-Bitfly commented on backend/pkg/commons/db2/database/bigtable.go:
```diff
@@ -0,0 +1,347 @@
+	if err != nil {
+		return fmt.Errorf("cannot ApplyBulk err: %w", err)
+	}
+	// TODO aggregate errs
```
```suggestion
```

Monika-Bitfly commented on backend/pkg/commons/db2/database/bigtable.go:
```diff
@@ -0,0 +1,347 @@
+// Close shuts down the BigTable by closing the Bigtable client connection
+// It returns an error if the operation fails
+func (b BigTable) Close() error {
+	if err := b.client.Close(); err != nil && status.Code(err) != codes.Canceled {
```
let's also check if the client is not nil before trying to close it

Monika-Bitfly commented on backend/pkg/commons/db2/database/bigtable_test.go:
```diff
@@ -0,0 +1,192 @@
+			}, {
+				key: "bar",
+			}},
+			expected: []string{"", "", "", ""},
```
is this correct expected output?

Tangui-Bitfly replied:
yes

Monika-Bitfly commented on backend/pkg/commons/db2/databasetest/bigtable.go:
```diff
@@ -0,0 +1,38 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	ctx := context.Background()
```
```suggestion
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
        defer cancel()
```

Tangui-Bitfly replied:
It's `databasetest/bigtable.go` is the sim bigtable but controlled through go. I don't really see the need for the context with timeout in that case.

Monika-Bitfly commented on backend/pkg/commons/db2/parser.go:
```diff
@@ -0,0 +1,195 @@
+}
+
+func (s *SenderFromDBSigner) ChainID() *big.Int {
+	panic("can't sign with SenderFromDBSigner")
```
```suggestion
	panic("can't sign with SenderFromDBSigner: ChainID method not supported")
```

Tangui-Bitfly replied:
I prefer to modify as little as possible when we "copy" code from somewhere.
The reasoning behind that is: let say in the future the source is update, we propagate this update, but we loose the local change.
So we may introduce a behavior that is prone to change.

Monika-Bitfly commented on backend/pkg/commons/db2/parser.go:
```diff
@@ -0,0 +1,195 @@
+	panic("can't sign with SenderFromDBSigner")
+}
+func (s *SenderFromDBSigner) Hash(tx *types.Transaction) common.Hash {
+	panic("can't sign with SenderFromDBSigner")
```
```suggestion
	panic("can't sign with SenderFromDBSigner: Hash method not supported")
```

Tangui-Bitfly replied:
same

Monika-Bitfly commented on backend/pkg/commons/db2/parser.go:
```diff
@@ -0,0 +1,195 @@
+	panic("can't sign with SenderFromDBSigner")
+}
+func (s *SenderFromDBSigner) SignatureValues(tx *types.Transaction, sig []byte) (R, S, V *big.Int, err error) {
+	panic("can't sign with SenderFromDBSigner")
```
```suggestion
	panic("can't sign with SenderFromDBSigner: SignatureValues method not supported")
```

Tangui-Bitfly replied:
same

Monika-Bitfly commented on backend/pkg/commons/db2/parser.go:
```diff
@@ -0,0 +1,195 @@
+// copy of senderFromServer
+// https://github.com/ethereum/go-ethereum/blob/v1.14.11/ethclient/signer.go#L30
+type SenderFromDBSigner struct {
+	addr      common.Address
```
```suggestion
	Address      common.Address
```

Tangui-Bitfly replied:
same

Monika-Bitfly commented on backend/pkg/commons/db2/parser.go:
```diff
@@ -0,0 +1,195 @@
+// https://github.com/ethereum/go-ethereum/blob/v1.14.11/ethclient/signer.go#L30
+type SenderFromDBSigner struct {
+	addr      common.Address
+	Blockhash common.Hash
```
```suggestion
	BlockHash common.Hash
```

Tangui-Bitfly replied:
same

Monika-Bitfly commented on backend/pkg/commons/db2/parser.go:
```diff
@@ -0,0 +1,195 @@
+	Input               string
+	Output              string
+	Error               string
+	RevertReason        string // todo have a look at this, it could improve revert message
```
```suggestion
	RevertReason        string // todo have a look at this, it could improve revert message
```
Is this comment still needed or can we remove it?

Tangui-Bitfly replied:
Yes it for future use. That field provide more context but is not present everywhere. If we change the indexer I want us to think about parsing this field.

Monika-Bitfly commented:
@Tangui-Bitfly Can you update the PR title to align with the template and to reflect the issue being addressed? Thanks! ^^

Monika-Bitfly commented on backend/pkg/commons/db2/raw/cache.go:
```diff
@@ -0,0 +1,119 @@
+	}
+	// TODO make warning not found in cache
+	block, err := c.store.ReadBlockByNumber(chainID, number)
+	if block != nil {
```
```suggestion
	if err != nil {
	    return nil, fmt.Errorf("block not found in cache: error %v", err)
	}
	if block != nil {
```

Tangui-Bitfly replied:
No you missed the point here. We don't want to handle the err, it's a cache layer, a passthrough. But we want to save the block in cache if it's not nil.

Monika-Bitfly commented on backend/pkg/commons/db2/raw/cache.go:
```diff
@@ -0,0 +1,119 @@
+		go c.unCacheBlockAfter(key, "", oneBlockTTL)
+		return v.(*FullBlockData), nil
+	}
+	// TODO make warning not found in cache
```
```suggestion
```

Tangui-Bitfly replied:
No once we have a proper logger I want to log something here

Monika-Bitfly commented on backend/pkg/commons/db2/raw/client.go:
```diff
@@ -0,0 +1,239 @@
+		request.Body = io.NopCloser(bytes.NewBuffer(body))
+	}()
+	var messages []*jsonrpc.Message
+	var isSingle bool
```
```suggestion
	var isSingleMsg bool
```
Maybe consider renaming this to `isSingleMsg` for better clarity

Monika-Bitfly commented on backend/pkg/commons/db2/raw/client_test.go:
```diff
@@ -0,0 +1,307 @@
+		block int64
+	}{
+		{
+			name:  "test block",
```
The test names could be more descriptive - eg this one we can consider renaming to "Test Block Retrieval" or something similar for better clarity

Tangui-Bitfly replied:
I prefer to keep it simple.
`TestBigTableClient.test_block` provide the same amount of info than `TestBigTableClient.Test_Block_Retrieval`
Or we could rename the test itself rather than the case.
But `TestBigTableClient` is describing what the test does, so I don't really how to improve it

Monika-Bitfly commented on backend/pkg/commons/db2/raw/client_test.go:
```diff
@@ -0,0 +1,307 @@
+		block FullBlockData
+	}{
+		{
+			name:  "test block",
```
Similar here, we can consider renaming it to "Test single block" and below to "Test single block with two uncles"

Tangui-Bitfly committed:
feat(test): add simulated backend for ethereum testing

Tangui-Bitfly committed:
feat(raw db): add raw store to interact with block raw table

db2/BigTableEthRaw: add chainID
db2/BigTableEthRaw: handle index for get uncle
db2/BigTableEthRaw: fix with empty tx
db2/BigTableEthRaw: add real condition test
db2/BigTableEthRaw: add WithFallback
db2/BigTableEthRaw: fallback for unsupported method
db2: add cache to raw store (#966)
db2: better cache + client + update store bigtable + clean up
db2/raw: add EthParse

Tangui-Bitfly commented:
> @Tangui-Bitfly Can you update the PR title to align with the template and to reflect the issue being addressed? Thanks! ^^

I'm not sure by what you want me to replace it with. There is no rule on PR name.

Monika-Bitfly commented:
@Tangui-Bitfly Could you please update the PR title to better reflect what this PR is implementing? If you're unsure, you can refer to the doc created by @peterbitfly or other open PRs in this repo eg. PR #1144 for guidance on how to name it. Thanks!

Tangui-Bitfly commented:
> @Tangui-Bitfly Could you please update the PR title to better reflect what this PR is implementing? If you're unsure, you can refer to the doc created by @peterbitfly or other open PRs in this repo eg. PR #1144 for guidance on how to name it. Thanks!

There is no guidance about PR name. Moreover PR are mostly irrelevant now because we don't have merge commit.

Monika-Bitfly commented:
> > @Tangui-Bitfly Could you please update the PR title to better reflect what this PR is implementing? If you're unsure, you can refer to the doc created by @peterbitfly or other open PRs in this repo eg. PR #1144 for guidance on how to name it. Thanks!
> 
> There is no guidance about PR name. Moreover PR are mostly irrelevant now because we don't have merge commit.

I’m not quite sure I understand how PRs are irrelevant now. A good PR title should clearly reflect the changes whether it’s a new feature or a fix. Right now the title doesn’t quite capture what this PR implements or addresses so it would be helpful if you could update it. This will make things clearer for everyone. Here are a couple of articles that can help guide you:
https://graphite.dev/blog/the-best-pr-title-of-all-time
https://blog.montrealanalytics.com/4-tips-for-effective-pull-request-naming-f60793998f04

Let me know if you have any questions, thanks!
